File: dotnet/Lexy/Lexy.Compiler/Language/AssignmentDefinition.cs
using System.Collections.Generic;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;

namespace Lexy.Compiler.Language
{
    public class AssignmentDefinition : Node
    {
        public Expression Expression { get; }
        public string Name { get; }

        public AssignmentDefinition(string name, Expression expression, SourceReference reference) : base(reference)
        {
            Expression = expression;
            Name = name;
        }

        public static AssignmentDefinition Parse(IParserContext context)
        {
            var valid = context.ValidateTokens<AssignmentDefinition>()
                .CountMinimum(3)
                .StringLiteral(0)
                .Operator(1, OperatorType.Assignment)
                .IsValid;

            if (!valid) return null;

            var line = context.CurrentLine;
            var name = line.Tokens.TokenValue(0);

            var value = ExpressionFactory.Parse(context.SourceCode.File, line.Tokens.TokensFrom(2), line);
            if (value.Status == ParseExpressionStatus.Failed)
            {
                context.Logger.Fail(context.LineStartReference(), value.ErrorMessage);
                return null;
            }

            var reference = context.LineStartReference();

            return new AssignmentDefinition(name, value.Expression, reference);
        }

        public override IEnumerable<INode> GetChildren()
        {
            yield return Expression;
        }

        protected override void Validate(IValidationContext context)
        {
            var expressionType = Expression.DeriveType(context);
            if (!context.FunctionCodeContext.Contains(Name))
            {
                context.Logger.Fail(Reference, $"Unknown variable '{Name}'.");
                return;
            }
            var variableType = context.FunctionCodeContext.GetVariableType(this.Name);
            if (!expressionType.Equals(variableType))
            {
                context.Logger.Fail(Reference, $"Variable '{Name}' of type '{variableType}' is not assignable from expression of type '{expressionType}'.");
            }
        }
    }
}

File: typescipt/src/language/scenarios/assignmentDefinition.ts


export class AssignmentDefinition extends Node {
   private readonly Expression valueExpression;
   private readonly Expression variableExpression;

   public ConstantValue ConstantValue
   public VariableReference Variable

   public VariableType VariableType { get; private set; }

   private AssignmentDefinition(VariableReference variable, ConstantValue constantValue, Expression variableExpression,
     Expression valueExpression, SourceReference reference)
     : base(reference) {
     Variable = variable;
     ConstantValue = constantValue;

     this.variableExpression = variableExpression;
     this.valueExpression = valueExpression;
   }

   public static parse(context: IParseLineContext): AssignmentDefinition {
     let line = context.Line;
     let tokens = line.Tokens;
     let reference = line.LineStartReference();

     let assignmentIndex = tokens.Find<OperatorToken>(token => token.Type == OperatorType.Assignment);
     if (assignmentIndex <= 0 || assignmentIndex == tokens.Length - 1) {
       context.Logger.Fail(reference, `Invalid assignment. Expected: 'Variable = Value'`);
       return null;
     }

     let targetExpression =
       ExpressionFactory.Parse(tokens.TokensFromStart(assignmentIndex), line);
     if (context.Failed(targetExpression, reference)) return null;

     let valueExpression =
       ExpressionFactory.Parse(tokens.TokensFrom(assignmentIndex + 1), line);
     if (context.Failed(valueExpression, reference)) return null;

     let variableReference = VariableReferenceParser.Parse(targetExpression.Result);
     if (context.Failed(variableReference, reference)) return null;

     let constantValue = ConstantValue.Parse(valueExpression.Result);
     if (context.Failed(constantValue, reference)) return null;

     return new AssignmentDefinition(variableReference.Result, constantValue.Result, targetExpression.Result,
       valueExpression.Result, reference);
   }

   public override getChildren(): Array<INode> {
     yield return variableExpression;
     yield return valueExpression;
   }

   protected override validate(context: IValidationContext): void {
     if (!context.VariableContext.Contains(Variable, context))
       //logger by IdentifierExpressionValidation
       return;

     let expressionType = valueExpression.DeriveType(context);

     VariableType = context.VariableContext.GetVariableType(Variable, context);
     if (expressionType != null && !expressionType.Equals(VariableType))
       context.Logger.Fail(Reference,
         $`Variable '{Variable}' of type '{VariableType}' is not assignable from expression of type '{expressionType}'.`);
   }
}
